### PR TITLE
feat(github): extract relation between pr and issue

### LINKED ISSUE #1257: Add a table to show relation between GitHub issue and PR
## Description
Add a table to show the relation between GitHub issue and PR, so that users can know the relation between these 2 entities.

## Describe the solution you'd like
It's a domain layer table, the relation is transformed by issue events like the pic below
1. This [issue](https://github.com/pingcap/tidb/issues/31467) is fixed by [PR 31731
](https://github.com/pingcap/tidb/pull/31731)
image

2. This [issue](https://github.com/pingcap/tidb/issues/31467) is mentioned in PR [31753](https://github.com/pingcap/tidb/pull/31753), [31754](https://github.com/pingcap/tidb/pull/31754), [31755](https://github.com/pingcap/tidb/pull/31755), etc.
image

3. Table design
- Table name: pull_request_issue
- 2 columns:
   - pull_request_id
   - issue_id

## Describe alternatives you've considered
The relation between PR and issue is only for Github

### LINKED ISSUE #1280: Github comments  and events' issueIds are numbers
## Describe the bug
In table.github_issue_comments and events, we stored issue number as issue id

## Expected behavior
We should store issue github id instead of number

### PR DESCRIPTION
**Pull request by warren830**
# Summary

add a table to show relation between pr and issue

### Key Points

- [ ] This is a breaking change
- [ ] New or existing documentation is updated

### Description
In tool layer, added a table github_pull_request_issues

### Does this close any open issues?
fix bugs caused by issue #1221
close #1280, #1256, #1060, #1257

### Screenshots
image

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

File: plugins/github/tasks/github_pull_request_issue_enricher.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+		}
+		for _, issueNumberStr := range issueNumberList {
+			issue := &githubModels.GithubIssue{}
+
+			issueNumberStr = strings.TrimSpace(issueNumberStr)
+			issueNumber, numFormatErr := strconv.Atoi(issueNumberStr)
+			if numFormatErr != nil {
+				continue
+			}
+			err = lakeModels.Db.Where("number = ?", issueNumber).Limit(1).Find(issue).Error
```
Comment (klesh) [inline plugins/github/tasks/github_pull_request_issue_enricher.go:95]: Number is not enough to identify a single issue on github, I think we need to add `repoId` as a composite identity

File: plugins/github/tasks/github_pull_request_issue_enricher.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+		//		return err
+		//	}
+		//}
+
+		if issueNumberListStr == "" {
+			continue
+		}
+		//replace https:// to #, then we can deal with later
+		if strings.Contains(issueNumberListStr, "https") {
+			httpsPrefixPattern := regexp.MustCompile(`https:\/\/github\.com\/\w+\/\w+issues\/`)
```
Comment (klesh) [inline plugins/github/tasks/github_pull_request_issue_enricher.go:74]: We should extract these kind of setting into configuration

File: plugins/github/tasks/github_pull_request_issue_enricher.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	//f, err := os.OpenFile("log.txt", os.O_CREATE|os.O_APPEND|os.O_RDWR, 0x666)
+	//defer f.Close()
+	//if err != nil {
+	//	return nil
+	//}
+	if len(prBodyClose) > 0 {
+		prBodyCloseRegex = regexp.MustCompile(prBodyClose)
+	} else {
+		patternStr := fmt.Sprintf(
+			`(?mi)^.*(fix|close|resolve|fixes|closes|resolves|fixed|closed|resolved)[\s]*(%s\/%s|issue)?(\s)?(((and ){0,1}(#|https:\/\/github\.com\/%s\/%s\/issues\/)\d+[ ]?)+)`,
```
Comment (klesh) [inline plugins/github/tasks/github_pull_request_issue_enricher.go:34]: We should extract these kind of setting into configuration

File: plugins/github/github.go [hunk 4]
```diff
@@ -73,27 +73,30 @@ func (plugin Github) Execute(options map[string]interface{}, progress chan<- flo
...
+			"convertRepos":                  true,
+			"convertIssues":                 true,
+			"convertIssueLabels":            true,
+			"convertPullRequests":           true,
+			"convertCommits":                true,
+			"convertPullRequestCommits":     true,
+			"convertPullRequestLabels":      true,
+			"convertNotes":                  true,
+			"convertUsers":                  true,
+			"createRefBugStats":             true,
```
Comment (klesh) [inline plugins/github/github.go:99]: if this task depends on Domain Layer Tables only, it should be placed into `github` plugin.
the github plugin normally get executed before domain layer tasks, maybe it should be moved into refdiff plugin

File: plugins/github/tasks/github_issue_comments_enricher.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package tasks
+
+import (
+	"context"
+	"github.com/merico-dev/lake/errors"
+	lakeModels "github.com/merico-dev/lake/models"
+	githubModels "github.com/merico-dev/lake/plugins/github/models"
+)
+
+func EnrichIssueComments(ctx context.Context, repoId int) error {
```
Comment (klesh) [inline plugins/github/tasks/github_issue_comments_enricher.go:10]: Combine EnrichPullRequestComments and EnrichIssueComments into EnrichComments

File: plugins/github/tasks/github_pull_request_issue_enricher.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+			return errors.TaskCanceled
+		default:
+		}
+		err = lakeModels.Db.ScanRows(cursor, githubPullRequst)
+		if err != nil {
+			return err
+		}
+
+		issueNumberListStr := getCloseIssueId(githubPullRequst.Body)
+		//record all matching string, for debug
```
Comment (klesh) [inline plugins/github/tasks/github_pull_request_issue_enricher.go:61]: remove debugging code

File: plugins/github/tasks/github_pull_request_issue_enricher.go (hunk 2 above)
Comment (warren830) [reply plugins/github/tasks/github_pull_request_issue_enricher.go:56]: This suits all repos as this is a standard link for issue.

**Review by klesh — COMMENTED** [review]
LGTM

**Review by klesh — APPROVED** [review]

File: plugins/github/tasks/github_comments_enricher.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			return err
+		}
+
+		issue := &githubModels.GithubIssue{}
+		err = lakeModels.Db.Where("number = ? and `repo_id` = ?", githubIssueComment.IssueNumber, repoId).Limit(1).Find(issue).Error
+		if err != nil {
+			return err
+		}
+		if issue.GithubId == 0 {
+			pr := &githubModels.GithubPullRequest{}
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/github/tasks/github_comments_enricher.go:40]: Yes, I added in the newest version

**Pull request merged**